`IStructure.__eq__` fails for integers

There are more edge cases like #2584 in `IStructure.__eq__`. E.g. comparing against `int`s also throws.

```py
from pymatgen.core import Lattice, Structure


coords = [[0.5, 0.5, 0.5], [0, 0, 0]]
species = ["Na", "Cl"]
lattice = Lattice.from_parameters(a=4, b=4, c=4, alpha=90, beta=90, gamma=90)
structure = Structure(lattice, species, coords)

structure == 42
```

```py
    def __eq__(self, other) -> bool:
        if other is self:
            return True
        if other is None:
            return False
>       if len(self) != len(other):
E       TypeError: object of type 'int' has no len()
```

I think a fix to handle all such edge cases would be to start with an `isinstance` check that returns `NotImplemented` if `other` is not a `(I)Structure`.

```py
def __eq__(self, other: object) -> bool:
    if not isinstance(other, IStructure):
        return NotImplemented
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -4,24 +4,24 @@ Change log
 v2022.7.19
 ----------
 
-This will be the final release with the pymatgen.analysis.defects 
-module included in the standard pymatgen package. This release will 
-include the older defects code by default, but can also be replaced with 
+This will be the final release with the pymatgen.analysis.defects
+module included in the standard pymatgen package. This release will
+include the older defects code by default, but can also be replaced with
 the newer defects code through installation of ``pymatgen-analysis-defects``.
 
-Subsequent versions of pymatgen will require 
-the additional installation of `pymatgen-analysis-defects <https://github.com/materialsproject/pymatgen-analysis-defects>`_ for all defect-related 
+Subsequent versions of pymatgen will require
+the additional installation of `pymatgen-analysis-defects <https://github.com/materialsproject/pymatgen-analysis-defects>`_ for all defect-related
 functionality via ``pip install pymatgen-analysis-defects``.
 
 Relevant imports will still be from the ``pymatgen.analysis.defects`` namespace but the code will now be maintained and developed in this separate repository.
 
-There will be significant changes to the defects code to support new functionality.Existing PyCDT users should use this version of pymatgen or older. Any questions 
+There will be significant changes to the defects code to support new functionality.Existing PyCDT users should use this version of pymatgen or older. Any questions
 about this change should be directed to Jimmy-Xuan Shen, @jmmshn.
 
-For more information about other pymatgen "add-on" packages, please see 
+For more information about other pymatgen "add-on" packages, please see
 `this page in our documentation <https://pymatgen.org/addons.html>`_.
 
-* Preparation for the removal of the defects module, PR #2582 by @jmmshn 
+* Preparation for the removal of the defects module, PR #2582 by @jmmshn
 
 v2022.7.8
 ---------

--- a/pymatgen/analysis/defects/core.py
+++ b/pymatgen/analysis/defects/core.py
@@ -25,7 +25,7 @@ warnings.warn(
     the new package on top of the old one so the entire `pymatgen.analysis.defects` namespace will be replaced.
 
     IMPORTANT: If you have already installed the new package and see this message it means the old package is
-    loaded first before the new package. See the documentation for more information, and raise an Issue on 
+    loaded first before the new package. See the documentation for more information, and raise an Issue on
     GitHub if this causes problems for your workflow.
     """,
     DeprecationWarning,

--- a/pymatgen/core/structure.py
+++ b/pymatgen/core/structure.py
@@ -1045,14 +1045,12 @@ class IStructure(SiteCollection, MSONable):
             return m.fit(self, other)
         return m.fit_anonymous(self, other)
 
-    def __eq__(self, other) -> bool:
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, IStructure):
+            return NotImplemented
         if other is self:
             return True
-        if other is None:
-            return False
         if len(self) != len(other):
-            return False
-        if not hasattr(other, "lattice"):
             return False
         if self.lattice != other.lattice:
             return False

--- a/pymatgen/core/structure.py
+++ b/pymatgen/core/structure.py
@@ -30,7 +30,9 @@ from typing import (
     Literal,
     Sequence,
     Set,
+    Sized,
     Union,
+    cast,
 )
 
 import numpy as np
@@ -1046,8 +1048,13 @@ class IStructure(SiteCollection, MSONable):
         return m.fit_anonymous(self, other)
 
     def __eq__(self, other: object) -> bool:
-        if not isinstance(other, IStructure):
+
+        # check for valid operand following class Student example from official functools docs
+        # https://docs.python.org/3/library/functools.html#functools.total_ordering
+        if not hasattr(other, "lattice") or not isinstance(other, Sized):
             return NotImplemented
+        other = cast(Structure, other)  # silence mypy errors below
+
         if other is self:
             return True
         if len(self) != len(other):

--- a/pymatgen/core/tests/test_structure.py
+++ b/pymatgen/core/tests/test_structure.py
@@ -86,6 +86,18 @@ class IStructureTest(PymatgenTest):
         self.assertEqual(df.attrs["Reduced Formula"], "Si")
         self.assertEqual(df.shape, (2, 8))
 
+    def test_equal(self):
+        struct = self.struct
+        self.assertTrue(struct == struct)
+        self.assertTrue(struct == struct.copy())
+        self.assertFalse(struct == 2 * struct)
+
+        self.assertFalse(struct == "a" * len(struct))  # GH-2584
+        self.assertFalse(struct is None)
+        self.assertFalse(struct == 42)  # GH-2587
+
+        self.assertTrue(struct == Structure.from_dict(struct.as_dict()))
+
     def test_matches(self):
         ss = self.struct * 2
         self.assertTrue(ss.matches(self.struct))
@@ -468,9 +480,9 @@ class IStructureTest(PymatgenTest):
         self.assertEqual(len(nn), 47)
         r = random.uniform(3, 6)
         all_nn = s.get_all_neighbors(r, True, True)
-        for i in range(len(s)):
-            self.assertEqual(4, len(all_nn[i][0]))
-            self.assertEqual(len(all_nn[i]), len(s.get_neighbors(s[i], r)))
+        for idx, site in enumerate(s):
+            self.assertEqual(4, len(all_nn[idx][0]))
+            self.assertEqual(len(all_nn[idx]), len(s.get_neighbors(site, r)))
 
         for site, nns in zip(s, all_nn):
             for nn in nns:
@@ -1411,7 +1423,7 @@ Sites (5)
 2 H     1.026719     0.000000    -0.363000
 3 H    -0.513360    -0.889165    -0.363000
 4 H    -0.513360     0.889165    -0.363000"""
-        self.assertEqual(self.mol.__str__(), ans)
+        self.assertEqual(str(self.mol), ans)
         ans = """Molecule Summary
 Site: C (0.0000, 0.0000, 0.0000)
 Site: H (0.0000, 0.0000, 1.0890)
@@ -1760,6 +1772,4 @@ class MoleculeTest(PymatgenTest):
 
 
 if __name__ == "__main__":
-    import unittest
-
     unittest.main()

--- a/pymatgen/core/tests/test_structure.py
+++ b/pymatgen/core/tests/test_structure.py
@@ -98,6 +98,11 @@ class IStructureTest(PymatgenTest):
 
         self.assertTrue(struct == Structure.from_dict(struct.as_dict()))
 
+        struct_2 = Structure.from_sites(struct)
+        self.assertTrue(struct, struct_2)
+        struct_2.apply_strain(0.5)
+        self.assertNotEqual(struct, struct_2)
+
     def test_matches(self):
         ss = self.struct * 2
         self.assertTrue(ss.matches(self.struct))
@@ -1251,14 +1256,6 @@ class StructureTest(PymatgenTest):
         self.structure.add_site_property("hello", [1, 2])
         s = Structure.from_sites(self.structure, to_unit_cell=True)
         self.assertEqual(s.site_properties["hello"][1], 2)
-
-    def test_magic(self):
-        s = Structure.from_sites(self.structure)
-        self.assertEqual(s, self.structure)
-        self.assertNotEqual(s, None)
-        s.apply_strain(0.5)
-        self.assertNotEqual(s, self.structure)
-        self.assertNotEqual(self.structure * 2, self.structure)
 
     def test_charge(self):
         s = Structure.from_sites(self.structure)

